File: tools/command/catchpoint.cpp
#include <fmt/base.h>
#include <fmt/format.h>

#include <libxdb/parse.hpp>
#include <libxdb/process.hpp>
#include <libxdb/syscalls.hpp>
#include <libxdb/types.hpp>
#include <sstream>
#include <unordered_set>

#include "handlers.hpp"

namespace {

std::vector<std::string> split_comma_separated(const std::string& str) {
    std::vector<std::string> result;
    std::stringstream ss(str);
    std::string item;

    while (std::getline(ss, item, ',')) {
        // Trim whitespace
        item.erase(0, item.find_first_not_of(" \t"));
        item.erase(item.find_last_not_of(" \t") + 1);
        if (!item.empty()) {
            result.push_back(item);
        }
    }

    return result;
}

std::optional<std::uint64_t> parse_syscall_identifier(const std::string& identifier) {
    // Try to parse as number first
    auto as_number = xdb::to_integral<std::uint64_t>(identifier);
    if (as_number) {
        return as_number;
    }

    // Try to parse as syscall name
    std::uint64_t syscall_id = xdb::syscall_name_to_id(identifier);
    if (syscall_id != static_cast<std::uint64_t>(-1)) {
        return syscall_id;
    }

    return std::nullopt;
}

}  // namespace

namespace xdb_handlers {

void handle_catchpoint_command(xdb::process& process, std::span<const std::string> args) {
    auto phelp = []() { print_help_init({"help", "catchpoint"}); };

    if (args.size() < 2) {
        phelp();
        return;
    }

    auto cmd = args[1];

    // Handle "catch syscall" or "catch sys" commands
    if (cmd == "syscall" || cmd == "sys") {
        if (args.size() < 3) {
            // Both "catch syscall" and "catch sys" with no arguments means catch all syscalls
            auto policy = xdb::syscall_catch_policy::catch_all();
            process.set_syscall_catch_policy(std::move(policy));
            fmt::println("Now catching all syscalls.");
            return;
        }

        const auto& syscall_arg = args[2];

        // Handle "catch sys none"
        if (syscall_arg == "none") {
            auto policy = xdb::syscall_catch_policy::catch_none();
            process.set_syscall_catch_policy(std::move(policy));
            fmt::println("No longer catching any syscalls.");
            return;
        }

        // Parse comma-separated syscall identifiers
        auto syscall_identifiers = split_comma_separated(syscall_arg);
        std::unordered_set<std::uint64_t> syscall_ids;

        for (const auto& identifier : syscall_identifiers) {
            auto syscall_id = parse_syscall_identifier(identifier);
            if (!syscall_id) {
                fmt::println("Error: Unknown syscall identifier '{}'", identifier);
                return;
            }
            syscall_ids.insert(*syscall_id);
        }

        if (syscall_ids.empty()) {
            fmt::println("Error: No valid syscall identifiers provided");
            return;
        }

        // Convert unordered_set to vector for syscall_catch_policy::catch_some
        std::vector<std::uint64_t> syscall_vector(syscall_ids.begin(), syscall_ids.end());
        auto policy = xdb::syscall_catch_policy::catch_some(syscall_vector);
        process.set_syscall_catch_policy(std::move(policy));

        fmt::println("Now catching {} syscall(s):", syscall_ids.size());
        for (auto id : syscall_ids) {
            auto name = xdb::syscall_id_to_name(id);
            fmt::println("  {} ({})", name, id);
        }

        return;
    }

    fmt::println("Unknown catchpoint command: {}", cmd);
    phelp();
}

}  // namespace xdb_handlers

File: src/bin/xdb/commands/catchpoint.rs
use std::collections::HashSet;

use xdb::syscalls::{syscall_id_to_name, syscall_name_to_id};
use xdb::{Process, SyscallCatchPolicy, parse};

pub fn handle(process: &mut Process, args: &[&str]) -> xdb::Result<()> {
    match args.get(1).copied() {
        // Handle "catch syscall" or "catch sys" commands
        Some("syscall" | "sys") => syscall(process, args),
        _ => {
            super::print_help(&["help", "catchpoint"]);
            Ok(())
        }
    }
}

fn syscall(process: &mut Process, args: &[&str]) -> xdb::Result<()> {
    match args.get(2).copied() {
        // Both "catch syscall" and "catch sys" with no arguments means catch all syscalls
        None => {
            process.set_syscall_catch_policy(SyscallCatchPolicy::All);
            println!("Now catching all syscalls.");
        }
        // Handle "catch sys none"
        Some("none") => {
            process.set_syscall_catch_policy(SyscallCatchPolicy::None);
            println!("No longer catching any syscalls.");
        }
        // Parse comma-separated syscall identifiers
        Some(list) => {
            let mut ids = HashSet::new();
            for item in list.split(',').map(str::trim).filter(|s| !s.is_empty()) {
                let Some(id) = resolve(item) else {
                    eprintln!("Unknown syscall identifier '{item}'");
                    return Ok(());
                };
                ids.insert(id);
            }
            if ids.is_empty() {
                eprintln!("No valid syscall identifiers provided");
                return Ok(());
            }
            println!("Now catching {} syscall(s):", ids.len());
            for id in &ids {
                let name = syscall_id_to_name(*id).unwrap_or("unknown");
                println!("  {name} ({id})");
            }
            process.set_syscall_catch_policy(SyscallCatchPolicy::Some(ids));
        }
    }
    Ok(())
}

fn resolve(item: &str) -> Option<u64> {
    // Try to parse as number first
    // Try to parse as syscall name
    parse::to_integral::<u64>(item, 10).or_else(|| syscall_name_to_id(item))
}
